### PR TITLE
Improve performance - Filter list of all tags first by those parsable to a semantic version

### PR DESCRIPTION
**Pull request by jkingry**
- reduces `Getting version tags from branch` time for our repository
  from a 100 seconds to 2.5 seconds

### CODE REVIEW

**Comment by jkingry** [comment]
I didn't see any existing performance tests, I've attempted a few times to write one using `System.Diagnostics.Stopwatch` but haven't yet succeeded. Are you open to using something like NBench or similar benchmark framework?

**Comment by asbjornu** [comment]
I haven't tried it, but I think [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) looks good. I'd happily accept pull requests with a few benchmark tests in it, especially if it's hooked up with some online service that can report performance degradation as Github checks to pull requests.

File: src/GitVersionCore/GitRepoMetadataProvider.cs [hunk 1]
```diff
@@ -34,19 +34,17 @@ public IEnumerable<SemanticVersion> GetVersionTagsOnBranch(Branch branch, string
...
             {
-                var tags = this.Repository.Tags.Select(t => t).ToList();
-
-                var versionTags = this.Repository.Commits.QueryBy(new CommitFilter
-                {
-                    IncludeReachableFrom = branch.Tip
-                })
-                .SelectMany(c => tags.Where(t => c.Sha == t.Target.Sha).SelectMany(t =>
+                var tags = new List<Tuple<Tag, SemanticVersion>>();
+                foreach(var t in this.Repository.Tags)
```
Comment (asbjornu) [inline src/GitVersionCore/GitRepoMetadataProvider.cs]: Can you please insert a space between `foreach` and `(`?

File: src/GitVersionCore/VersionCalculation/BaseVersionCalculators/TaggedCommitVersionStrategy.cs [hunk 2]
```diff
@@ -19,21 +19,31 @@ public override IEnumerable<BaseVersion> GetVersions(GitVersionContext context)
 
         public IEnumerable<BaseVersion> GetTaggedVersions(GitVersionContext context, Branch currentBranch, DateTimeOffset? olderThan)
         {
-            var allTags = context.Repository.Tags
-                .Where(tag => !olderThan.HasValue || ((Commit) tag.PeeledTarget()).When() <= olderThan.Value)
-                .ToList();
+            var allTags = new List<Tuple<Tag, SemanticVersion>>();
+            foreach(var tag in context.Repository.Tags)
```
Comment (asbjornu) [inline src/GitVersionCore/VersionCalculation/BaseVersionCalculators/TaggedCommitVersionStrategy.cs]: Can you please insert a space between `foreach` and `(`?

File: src/GitVersionCore/GitRepoMetadataProvider.cs [hunk 3]
```diff
@@ -34,19 +34,17 @@ public IEnumerable<SemanticVersion> GetVersionTagsOnBranch(Branch branch, string
...
             using (Logger.IndentLog(string.Format("Getting version tags from branch '{0}'.", branch.CanonicalName)))
             {
-                var tags = this.Repository.Tags.Select(t => t).ToList();
-
-                var versionTags = this.Repository.Commits.QueryBy(new CommitFilter
-                {
-                    IncludeReachableFrom = branch.Tip
-                })
-                .SelectMany(c => tags.Where(t => c.Sha == t.Target.Sha).SelectMany(t =>
+                var tags = new List<Tuple<Tag, SemanticVersion>>();
```
Comment (asbjornu) [inline src/GitVersionCore/GitRepoMetadataProvider.cs]: Can you please extract this tag filter logic out to a descriptive method?

**Comment by asbjornu** [comment]
Seeing how this might fix #1361, I'll ignore what I've written in the review and just go ahead and merge. A followup PR that addresses the comments would be appreciated, though. Thanks for this!

**Pull request merged**